src/Init/System/Promise.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+prelude
+import Init.System.IO
+
+namespace IO
+
+/-- Internally, a `Promise` is just a `Task` that is in the "Promised" or "Finished" state. -/
+private opaque PromiseImpl (α : Type) : { P : Type // Nonempty α ↔ Nonempty P } :=
+  ⟨Task α, fun ⟨_⟩ => ⟨⟨‹_›⟩⟩, fun ⟨⟨_⟩⟩ => ⟨‹_›⟩⟩
+
+/--
+`Promise α` allows you to create a `Task α` whose value is provided later by calling `resolve`.
+
+Typical usage is as follows:
+1. `let promise ← Promise.new` creates a promise
+2. `promise.result : Task α` can now be passed around
+3. `promise.result.get` blocks until the promise is resolved
+4. `promise.resolve a` resolves the promise
+5. `promise.result.get` now returns `a`
+
+Every promise must eventually be resolved.
+Otherwise the memory used for the promise will be leaked,
```

Comment:
Couldn't we skip the `Deactivated` state in this case? We would need more data in the `lean_task_impl` so that `deactivate_task` can identify this case I suppose.

Reply:
Indeed, we'd need more data.  Right now, a Promised task looks the same as a Running task.  And Running tasks are owned by the worker (and thus can't be freed).